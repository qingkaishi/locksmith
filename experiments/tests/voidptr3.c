#include <pthread.h>

struct foo {
  int u;
};

struct bar {
  float v;  /* should be a race on this */
  int m;    /* but no race on this */
};

void *dostuff1(void *x) {
  struct foo *z = (struct foo *) x;
  z->u = 5;
}

void *dostuff2(void *y) {
  struct bar *w = (struct bar *) y;
  w->v = 5;
  w->m = 3;
}

void *dostuff3(void *z) {
  struct bar *t = (struct bar *) z;
  t->v = 6;
}

void *dostuff4(void *s) {
  int *r = (int *) s;
  *r = 3;
}

void *dostuff5(void *q) {
  int *p = (int *) q;
  *p = 4;
}

int main(int argc, char **argv) {
  struct foo a;
  struct bar b;
  int c;
  pthread_t t1;
  pthread_create(&t1, NULL, dostuff1, &a);
  pthread_create(&t1, NULL, dostuff2, &b);
  pthread_create(&t1, NULL, dostuff3, &b);
  pthread_create(&t1, NULL, dostuff4, &c);
  pthread_create(&t1, NULL, dostuff5, &c);
}
