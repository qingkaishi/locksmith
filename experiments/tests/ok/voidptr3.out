************************* STARTING *************************

shared: &c:voidptr3.c:41
        alloc.v:voidptr3.c:40

Warning: Possible data race: &c:voidptr3.c:41 is not protected!
 references:
  dereference of r:voidptr3.c:29 at voidptr3.c:30
    &c:voidptr3.c:41 => cast:voidptr3.c:29 at dostuff4 voidptr3.c:46
                     => r:voidptr3.c:29
  locks acquired:
    <empty>
  in: main at voidptr3.c:38 

  dereference of p:voidptr3.c:34 at voidptr3.c:35
    &c:voidptr3.c:41 => cast:voidptr3.c:34 at dostuff5 voidptr3.c:47
                     => p:voidptr3.c:34
  locks acquired:
    <empty>
  in: main at voidptr3.c:38 



Warning: Possible data race: alloc.v:voidptr3.c:40 is not protected!
 references:
  dereference of *w.v:voidptr3.c:18 at voidptr3.c:19
    alloc.v:voidptr3.c:40 => b.v:voidptr3.c:40
                          => *w.v:voidptr3.c:18 at dostuff2 voidptr3.c:44
  locks acquired:
    <empty>
  in: main at voidptr3.c:38 

  dereference of *t.v:voidptr3.c:24 at voidptr3.c:25
    alloc.v:voidptr3.c:40 => b.v:voidptr3.c:40
                          => *t.v:voidptr3.c:24 at dostuff3 voidptr3.c:45
  locks acquired:
    <empty>
  in: main at voidptr3.c:38 


*************************** DONE ***************************
